File: 2022/src/day02.rs
use crate::helpers;

pub fn solve() {
    let lines = helpers::lines_from_file("./input/day_02.txt");
    println!("Day 02:");
    println!("{}", part_1(&lines));
    println!("{}", part_2(&lines));
}

fn part_1(lines: &Vec<String>) -> u32 {

    let lookup = [[4, 8, 3], [1, 5, 9], [7, 2, 6]];
    let mut total_score = 0;

    for line in lines {
        let his = line.as_bytes()[0] - b'A';
        let mine = line.as_bytes()[2] - b'X';

        total_score += lookup[his as usize][mine as usize];
    }

    return total_score;
}

fn part_2(lines: &Vec<String>) -> u32 {
    let lookup = [[3, 4, 8], [1, 5, 9], [2, 6, 7]];
    let mut total_score = 0;

    for line in lines {
        let his = line.as_bytes()[0] - b'A';
        let mine = line.as_bytes()[2] - b'X';

        total_score += lookup[his as usize][mine as usize];
    }

    return total_score;
}

File: 2022/src/day02.cpp
#include <string>
#include <sstream>
#include <fstream>
#include <vector>
#include <iostream>
#include <algorithm>

#include "common.h"

int part_1(std::vector<std::string>& lines)
{
    std::vector<std::vector<int>> lookup = {
        {4, 8, 3}, {1, 5, 9}, {7, 2, 6}
    };
    int total_score = 0;

    for (auto line: lines) {
        int his  = line[0] - 'A';
        int mine = line[2] - 'X';

        total_score += lookup[his][mine];
    }

    return total_score;
}

int part_2(std::vector<std::string>& lines)
{
    std::vector<std::vector<int>> lookup = {
        {3, 4, 8}, {1, 5, 9}, {2, 6, 7}
    };
    int total_score = 0;

    for (auto line: lines) {
        int his  = line[0] - 'A';
        int mine = line[2] - 'X';

        total_score += lookup[his][mine];
    }

    return total_score;
}

void day02(std::string& input_path) {
    std::cout << "Day 2:" << std::endl;

    auto lines = get_input(input_path);
    std::cout << part_1(lines) << std::endl;
    std::cout << part_2(lines) << std::endl;
}
